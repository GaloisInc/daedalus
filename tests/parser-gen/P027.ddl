
def Frac =
       { @dot = Match "."
       ; @t = { @ds = Many Digit; ^ numBase 10 ds }
       ; ^ t
       }

def NoFrac = { @ds = Match "" ; ^ -42 }

def Continue = ( Frac <| NoFrac )

def UnsignedNumber = Choose{
                   { n = Natural
                   ; y = Continue
                   }
                 ;
                   { @n = Match ""
                   ; n = ^ 0
                   ; @x = Match1 '.'
                   ; y = Natural
                   }
                 }

def numBase base ds       = for (val = 0; d in ds) (val * base + d)

def Natural = { @ds = Many (1..) Digit; ^ numBase 10 ds }

def Digit     = { @d = Match1 ('0' .. '9'); ^ d - '0' as int }

def Main = { x = Many { UnsignedNumber; Match " "} ;  }
