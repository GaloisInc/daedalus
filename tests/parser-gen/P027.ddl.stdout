[WARNING] ./P027.ddl:12:22--22:18:
            Using unbiased choice may be inefficient.
module P027
 
--- Imports:
 
--- Type defs:
type P027.UnsignedNumber = { n: int
                           ; y: int
                           }
 
type P027.Main = { x: [[uint 8]]
                 }
 
--- Rules:
 
P027.numBase ?a0 ?a1 ?a2 (Literal 0 ?a2) (HasKey ?a0 ?a1) (HasElement ?a0 ?a2) (Traversable ?a0) (Arith ?a2) (base : ?a2) (ds : ?a0) : ?a2 =
  for ((val : ?a2) = 0; (d : ?a2) in ds)
    ((val * base) + d)
 
P027.Digit : Grammar int =
  do (d : uint 8) <- Match ('0' .. '9')
     ($$ : int) <- pure (d - '0' as int)
     pure $$
 
P027.Frac : Grammar int =
  do @MatchBytes "."
     (t : int) <- do (ds : [int]) <- Many[] P027.Digit
                     ($$ : int) <- pure (P027.numBase [int] (uint 64) int 10 ds)
                     pure $$
     ($$ : int) <- pure t
     pure $$
 
P027.NoFrac ?a0 (Literal -42 ?a0) : Grammar ?a0 =
  do @MatchBytes ""
     ($$ : ?a0) <- pure -42
     pure $$
 
P027.Continue : Grammar int =
  Choose biased
    { P027.Frac
    | P027.NoFrac int
    }
 
P027.Natural : Grammar int =
  do (ds : [int]) <- Many[ 1 .. ] P027.Digit
     ($$ : int) <- pure (P027.numBase [int] (uint 64) int 10 ds)
     pure $$
 
P027.UnsignedNumber : Grammar P027.UnsignedNumber =
  Choose fair
    { do (n : int) <- P027.Natural
         (y : int) <- P027.Continue
         pure {n = n,
               y = y}
    | do @MatchBytes ""
         (n : int) <- pure 0
         @Match {'.'}
         (y : int) <- P027.Natural
         pure {n = n,
               y = y}
    }
 
P027._Digit : Grammar {} =
  @Match ('0' .. '9')
 
P027._Frac : Grammar {} =
  do @MatchBytes "."
     @Many[] P027._Digit
 
P027._NoFrac ?a0 (Literal -42 ?a0) : Grammar {} =
  @MatchBytes ""
 
P027._Continue : Grammar {} =
  Choose biased
    { P027._Frac
    | P027._NoFrac int
    }
 
P027._Natural : Grammar {} =
  @Many[ 1 .. ] P027._Digit
 
P027._UnsignedNumber : Grammar {} =
  Choose fair
    { do P027._Natural
         P027._Continue
    | do @MatchBytes ""
         @Match {'.'}
         P027._Natural
    }
 
P027.Main : Grammar P027.Main =
  do (x : [[uint 8]]) <- Many[] do P027._UnsignedNumber
                                   ($$ : [uint 8]) <- MatchBytes " "
                                   pure $$
     pure {x = x}
 
P027._Main : Grammar {} =
  @Many[] do P027._UnsignedNumber
             @MatchBytes " "
